File: src/conventions/pipeline/strategy-evaluation.ts
import type { BiddingStrategy } from "../core/strategy-types";
import type { MachineRegisters } from "../core/module-surface";
import type { PipelineResult } from "./pipeline-types";
import type { HandoffTrace } from "./evaluation/provenance";
import type { TeachingProjection } from "../teaching/teaching-types";
import type { PracticalRecommendation, PosteriorSummary } from "../core/strategy-types";
import type { SurfaceGroup } from "../teaching/teaching-types";
import type { ExplanationCatalog } from "../core/explanation-catalog";
import type { EvaluatedFacts } from "../core/fact-catalog";
import type { AuctionContext } from "../core/committed-step";

/** Lightweight DTO for convention machine state — avoids importing from conventions/core. */
export interface MachineDebugSnapshot {
  readonly currentStateId: string;
  readonly stateHistory: readonly string[];
  readonly transitionHistory: readonly string[];
  readonly activeSurfaceGroupIds: readonly string[];
  readonly registers: MachineRegisters;
  readonly diagnostics: readonly { readonly level: string; readonly message: string; readonly moduleId?: string }[];
  readonly handoffTraces: readonly HandoffTrace[];
  readonly submachineStack: readonly { readonly parentMachineId: string; readonly returnStateId: string }[];
}

/** Unified evaluation snapshot — all pipeline outputs from the most recent suggest() call. */
export interface StrategyEvaluation {
  /** Practical recommendation (what an experienced player might prefer). Null when not produced. */
  readonly practicalRecommendation: PracticalRecommendation | null;
  /** Convention-level surface groups for relationship-aware grading. Null if not configured. */
  readonly surfaceGroups: readonly SurfaceGroup[] | null;
  /** Full pipeline result from the meaning-pipeline evaluation. Null when not produced. */
  readonly pipelineResult: PipelineResult | null;
  /** Posterior summary. Null when posterior engine not wired. */
  readonly posteriorSummary: PosteriorSummary | null;
  /** Explanation catalog for enriching teaching projections. Null when not available. */
  readonly explanationCatalog: ExplanationCatalog | null;
  /** Teaching projection from the meaning-pipeline evaluation. Null when not produced. */
  readonly teachingProjection: TeachingProjection | null;
  /** Evaluated facts from the pipeline evaluation. Null before first evaluation. */
  readonly facts: EvaluatedFacts | null;
  /** Machine/protocol state from the evaluation. Null when no machine is wired. */
  readonly machineSnapshot: MachineDebugSnapshot | null;
  /** Observation log and snapshot for rule-based surface selection. Null when not produced. */
  readonly auctionContext: AuctionContext | null;
}

/** Extended strategy interface for convention-based strategies that produce practical recommendations.
 *  Returns the full evaluation snapshot from the most recent suggest() call (null before first suggest()). */
export interface ConventionStrategy extends BiddingStrategy {
  getLastEvaluation(): StrategyEvaluation | null;
}

File: src-tauri/crates/bridge-conventions/src/adapter/strategy_evaluation.rs
//! Strategy evaluation — unified debug container.
//!
//! Mirrors TS from `pipeline/strategy-evaluation.ts`.

use serde::{Deserialize, Serialize};

use crate::fact_dsl::types::EvaluatedFacts;
use crate::pipeline::evaluation::provenance::HandoffTrace;
use crate::pipeline::observation::committed_step::AuctionContext;
use crate::pipeline::types::PipelineResult;
use crate::teaching::teaching_types::{SurfaceGroup, TeachingProjection};

/// Lightweight DTO for convention machine state.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MachineDebugSnapshot {
    /// Current state ID (populated from rules path).
    pub current_state_id: String,
    /// Active surface group IDs (populated from rules path).
    pub active_surface_group_ids: Vec<String>,
    /// Diagnostic messages (populated from rules path).
    pub diagnostics: Vec<DiagnosticEntry>,
    /// State history (legacy FSM — None from rules path).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub state_history: Option<Vec<String>>,
    /// Transition history (legacy FSM — None from rules path).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub transition_history: Option<Vec<String>>,
    /// Handoff traces (legacy FSM — None from rules path).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub handoff_traces: Option<Vec<HandoffTrace>>,
    /// Submachine stack (legacy FSM — None from rules path).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub submachine_stack: Option<Vec<SubmachineEntry>>,
    /// Machine registers (legacy FSM — None from rules path).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub registers: Option<serde_json::Value>,
}

/// A diagnostic entry.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DiagnosticEntry {
    pub level: String,
    pub message: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub module_id: Option<String>,
}

/// Submachine stack entry.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SubmachineEntry {
    pub parent_machine_id: String,
    pub return_state_id: String,
}

/// Unified evaluation snapshot — all pipeline outputs from the most recent suggest() call.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StrategyEvaluation {
    /// Practical recommendation. Null when not produced.
    pub practical_recommendation: Option<PracticalRecommendation>,
    /// Convention-level surface groups. Null if not configured.
    pub surface_groups: Option<Vec<SurfaceGroup>>,
    /// Full pipeline result. Null when not produced.
    pub pipeline_result: Option<PipelineResult>,
    /// Posterior summary. Null when posterior engine not wired.
    pub posterior_summary: Option<serde_json::Value>,
    /// Explanation catalog — module-level explanation entries.
    pub explanation_catalog: Option<ExplanationCatalog>,
    /// Teaching projection. Null when not produced.
    pub teaching_projection: Option<TeachingProjection>,
    /// Evaluated facts. Null before first evaluation.
    pub facts: Option<EvaluatedFacts>,
    /// Machine/protocol state. Null when no machine is wired.
    pub machine_snapshot: Option<MachineDebugSnapshot>,
    /// Auction context. Null when not produced.
    pub auction_context: Option<AuctionContext>,
}

/// Explanation catalog — typed collection of module-level explanation entries.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ExplanationCatalog {
    pub entries: Vec<ExplanationCatalogEntry>,
}

/// A single explanation catalog entry.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ExplanationCatalogEntry {
    pub module_id: String,
    pub surface_count: usize,
    pub explanation: String,
}

/// Practical recommendation — what an experienced player might prefer.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PracticalRecommendation {
    pub call: bridge_engine::types::Call,
    pub reason: String,
    pub confidence: f64,
}
